shell/platform/android/io/flutter/plugin/editing/TextEditingDelta.java

```diff
@@ -0,0 +1,200 @@
+    final boolean isInsertingInsideComposingRegion = start + tbend > end;
+
+    // To consider the cases when autocorrect increases the length of the text being composed by
+    // one, but changes more than one character.
+    final boolean isOriginalComposingRegionTextChanged =
+        (isCalledFromDelete || isDeletingInsideComposingRegion || isReplacedByShorter)
+            || !oldEditable
+                .subSequence(start, end)
+                .toString()
+                .equals(tb.subSequence(tbstart, end - start).toString());
+
+    final boolean isEqual = oldEditable.equals(newEditable);
+
+    // A replacement means the original composing region has changed, anything else will be
+    // considered an insertion.
+    final boolean isReplaced =
+        isOriginalComposingRegionTextChanged
+            && (isReplacedByLonger || isReplacedBySame || isReplacedByShorter);
+
+    if (isEqual) {
+      Log.v(TAG, "A TextEditingDelta for an TextEditingDeltaType.equality has been created.");
+      setDeltas(oldEditable, "", "TextEditingDeltaType.equality", -1, -1);
+    } else if (isCalledFromDelete || isDeletingInsideComposingRegion) {
+      Log.v(TAG, "A TextEditingDelta for a TextEditingDeltaType.deletion has been created.");
+      final int startOfDelete;
+      if (isDeletionGreaterThanOne) {
+        startOfDelete = start;
+      } else {
+        startOfDelete = end - 1;
+      }
+
+      setDeltas(
+          oldEditable,
+          oldEditable.subSequence(start + tbend, end).toString(),
+          "TextEditingDeltaType.deletion",
+          startOfDelete,
+          end);
+    } else if ((start == end || isInsertingInsideComposingRegion)
+        && !isOriginalComposingRegionTextChanged) {
+      Log.v(TAG, "A TextEditingDelta for an TextEditingDeltaType.insertion has been created.");
+      setDeltas(
+          oldEditable,
+          tb.subSequence(end - start, tbend).toString(),
+          "TextEditingDeltaType.insertion",
+          end,
+          end);
+    } else if (isReplaced) {
+      Log.v(TAG, "A TextEditingDelta for a TextEditingDeltaType.replacement has been created.");
+      setDeltas(
+          oldEditable,
+          tb.subSequence(tbstart, tbend).toString(),
+          "TextEditingDeltaType.replacement",
+          start,
+          end);
+    }
+  }
+
+  @VisibleForTesting
+  public TextEditingDelta(
+      CharSequence oldText,
+      CharSequence deltaText,
+      CharSequence deltaType,
+      int deltaStart,
+      int deltaEnd,
+      int selectionStart,
+      int selectionEnd,
+      int composingStart,
+      int composingEnd) {
+    newSelectionStart = selectionStart;
+    newSelectionEnd = selectionEnd;
+    newComposingStart = composingStart;
+    newComposingEnd = composingEnd;
+
+    setDeltas(oldText, deltaText, deltaType, deltaStart, deltaEnd);
+  }
+
+  public void setNewComposingStart(int newStart) {
```

Comment:
Do you still need to expose these methods? Would it be possible to make this class immutable?

Reply:
I think this is possible now.